include/boost/geometry/algorithms/detail/overlay/discard_duplicate_turns.hpp

```diff
@@ -88,72 +88,154 @@ inline bool corresponding_turn(Turn const& turn, Turn const& start_turn,
         }
-        index++;
     }
+}
 
-    // 2: Verify all other turns if they are present in the map, and if so,
-    //    if they have the other turns as corresponding
-    for (auto const& turn : turns)
+//  Discard turns for the following (rare) case:
+//  - they are consecutive
+//  - the first has a touch, the second a touch_interior
+// And then one of the segments touches the others next in the middle.
+// This is geometrically not possible, and caused by floating point precision.
+// Discard the second (touch interior)
+template <typename Turns, typename Geometry0, typename Geometry1>
+void discard_touch_touch_interior_turns(Turns& turns,
+                                        Geometry0 const& geometry0,
+                                        Geometry1 const& geometry1)
+{
+    for (auto& current_turn : turns)
     {
-        // Any turn which "crosses" does not have a corresponding turn.
-        // Also avoid comparing "start" with itself.
-        if (turn.method != method_crosses && turn.method != method_start)
+        if (current_turn.method != method_touch_interior)
         {
-            for (auto const& op : turn.operations)
+            // Because touch_interior is a rarer case, it is more efficient to start with that
+            continue;
+        }
+        for (auto const& previous_turn : turns)
+        {
+            if (previous_turn.method != method_touch)
             {
-                auto it = start_turns_per_segment.find(adapt_id(op.seg_id));
-                if (it != start_turns_per_segment.end())
-                {
-                    for (std::size_t const& i : it->second)
-                    {
-                        if (turns[i].cluster_id != turn.cluster_id)
-                        {
-                            // The turns are not part of the same cluster,
-                            // or one is clustered and the other is not.
-                            // This is not corresponding.
-                            continue;
-                        }
-                        if (corresponding_turn(turn, turns[i],
-                                               geometry0, geometry1))
-                        {
-                            turns[i].discarded = true;
-                        }
-                    }
-                }
+                continue;
+            }
+
+            // Compare 0 with 0 and 1 with 1
+            // Note that 0 has always source 0 and 1 has always source 1
+            // (not in buffer). Therefore this comparison is OK.
+            // MAYBE we need to check for buffer.
+            bool const common0 = current_turn.operations[0].seg_id == previous_turn.operations[0].seg_id;
+            bool const common1 = current_turn.operations[1].seg_id == previous_turn.operations[1].seg_id;
+
+            // If one of the operations is common, and the other is not, then there is one comment segment.
+            bool const has_common_segment = common0 != common1;
```

Comment:
I’ll adapt it slightly tomorrow.
Also this change somehow introduces a dependency.
```
gcc.compile.c++ bin.v2/libs/geometry/test/headers/h-algorithms-detail-overlay-discard_duplicate_turns.test/gcc-11/debug/x86_64/threading-multi/visibility-hidden/h-algorithms-detail-overlay-discard_duplicate_turns.o
```